runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -9240,6 +9240,17 @@ inlineNanoTime(
       // result = reg + result
       generateRegMemInstruction(TR::InstOpCode::LEA8RegMem, node, result, generateX86MemoryReference(reg, result, 0, cg), cg);
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+      if (fej9->inSnapshotMode())
```

Comment:
This is not a correct check to determine if nano time adjustment is required.
https://github.com/eclipse-openj9/openj9/blob/a57e6d99bb5fbb7ad12946a3aa7748dbd94f4945/runtime/compiler/env/VMJ9.cpp#L9345-L9348
`isCheckpointAllowed()` is `TRUE` before a checkpoint taken in which `nanoTimeMonotonicClockDelta` is zero and no adjustment required though it won't hurt to do so.
However `isCheckpointAllowed()` is `FALSE` after a checkpoint taken at `CRIURestoreNonPortableMode` which is a main mode in use, though a checkpoint is not allowed `nanoTimeMonotonicClockDelta` is not zero after a checkpoint, and adjustment is needed.
I think you could always apply `result = result - nanoTimeMonotonicClockDelta`, or retrieve `nanoTimeMonotonicClockDelta` value first, only do the subtraction if it is non-zero.

Reply:
Thanks, you're right. I assumed it was true on both sides of the C/R.

I'm going to rework the JIT queries to add one that uses `isCRIUSupportEnabled` (and make it clearer as to what question each query is answering) instead of looking at the the value of `nanoTimeMonotonicClockDelta` because we'll likely have other cases where we need to generate C/R-aware code after restore.

I modified the tests to have separate test cases for pre and post C/R compilation and the post one catches this bug.